Random

mgautierfr opened the pull request:
Need merge of ~~#466~~ #487 first.

Fixes #488

kelson42 commented:
@mgautierfr will that fix https://github.com/openzim/warc2zim/issues/37 ?

veloman-yunkan commented on src/archive.cpp:
```diff
@@ -178,6 +178,23 @@ namespace zim
+      auto idx = ((double)rand() / ((double)RAND_MAX+1)) * frontEntryCount;
+      auto entry = getEntryByTitle(idx);
+      auto item = entry.getItem(true);
+      std::cerr << item.getPath() << "=>" << item.getMimetype() << std::endl;
```
Do we need this bare `std::cerr` here? Maybe wrap it in a logging statement?

mgautierfr replied:
I've removed the line.

veloman-yunkan commented on test/archive.cpp:
```diff
@@ -162,6 +162,25 @@ TEST(ZimArchive, openRealZimArchive)
+    "wikipedia_en_climate_change_nopic_2020-01.zim"
+  };
+
+  for ( const std::string fname : zimfiles ) {
+    const auto path = zim::DEFAULTFS::join("data", fname);
+    const TestContext ctx{ {"path", path } };
+    const zim::Archive archive(path);
+    const auto randomEntry = archive.getRandomEntry();
+    const auto item = randomEntry.getItem(true);
+    ASSERT_TRUE(item.getMimetype().find("text/html") != std::string::npos) << ctx;
+  }
```
It also makes sense to call `getRandomEntry()` several times and check that its result is indeed random. To this end `small.zim` will need to be either excluded from this test, or enhanced with more pages (currently it has only one front page).

mgautierfr replied:
Done

veloman-yunkan reviewed (changes requested):
I reviewed only the last commit. In fact I wonder, why this PR doesn't have base branch set to `indexes`. Then only the last commit will show up in this PR and the base branch will be automatically reset to `master` when `indexes` is merged.

kelson42 commented:
@mgautierfr Would be an good opportunity to fix #488 here IMO.

mgautierfr commented:
The last push is a better implementation of random :
- Use cpp random version
- Make the random thread safe (rand is not)
- Properly initialize random with time as seed (should fix #488)

veloman-yunkan commented on src/archive.cpp:
```diff
@@ -184,6 +184,22 @@ namespace zim
+      }
+      return entry;
+    }
+    throw EntryNotFound("Cannot found valid random entry");
```
found → find

veloman-yunkan commented on include/zim/archive.h:
```diff
@@ -212,6 +212,16 @@ namespace zim
+       * The entry is pick randomly from the front artice list.
+       *
+       * @return A random entry.
+       * @exception EntryNoFound If no valid random entry can be found.
```
EntryNoFound → EntryNotFound

veloman-yunkan commented on src/fileimpl.h:
```diff
@@ -111,7 +111,11 @@ namespace zim
 
       entry_index_t getStartUserEntry() const { return m_startUserEntry; }
       entry_index_t getEndUserEntry() const { return m_endUserEntry; }
+      // The number of entry added by the creator. (So excluding index, ...).
+      // On new namespace scheme, number of entry in C namespace
```
entry → entries

veloman-yunkan commented on src/fileimpl.h:
```diff
@@ -111,7 +111,11 @@ namespace zim
+      // The number of entry added by the creator. (So excluding index, ...).
+      // On new namespace scheme, number of entry in C namespace
       entry_index_t getUserEntryCount() const { return m_endUserEntry - m_startUserEntry; }
+      // The number of entry that can be considered as front article (no resource)
```
entry → entries

veloman-yunkan commented on src/tools.cpp:
```diff
@@ -114,3 +116,13 @@ std::tuple<char, std::string> zim::parseLongPath(const std::string& longPath)
+  static std::mutex mutex;
+
+  std::lock_guard<std::mutex> l(mutex);
+  return ((double)random() / RAND_MAX) * max;
```
use `random.max()` instead of `RAND_MAX`

veloman-yunkan commented on src/tools.h:
```diff
@@ -34,6 +34,12 @@ namespace zim {
 
   std::tuple<char, std::string> parseLongPath(const std::string& longPath);
+
+  /** Return a random number between [0, max]
```
between → from the range

veloman-yunkan commented on src/archive.cpp:
```diff
@@ -184,6 +184,22 @@ namespace zim
+    auto frontEntryCount = m_impl->getFrontEntryCount().v;
+    int watchdog = 42;
+    while(--watchdog) {
+      auto idx = randomNumber(frontEntryCount);
```
`frontEntryCount` → `frontEntryCount-1` (and handle the case of an empty ZIM archive)

mgautierfr replied:
Good catch

veloman-yunkan commented on test/archive.cpp:
```diff
@@ -153,6 +153,33 @@ TEST(ZimArchive, openRealZimArchive)
+    }
+    for (auto i = 0; i<10; i++) {
+      const auto newRandomEntry = archive.getRandomEntry();
+      ASSERT_NE(randomEntry.getIndex(), newRandomEntry.getIndex());
```
This can occasionally fail

veloman-yunkan commented on test/random.cpp:
```diff
@@ -0,0 +1,52 @@
+
+TEST(Random, pseudoUnique)
+{
+  auto prev = randomNumber(1000000);
+  for (auto i=0; i<1000; i++) {
+    auto r = randomNumber(1000000);
+    ASSERT_NE(prev, r);
+    prev = r;
+  }
```
An alternating sequence 0, 1, 0, 1, ... returned by `randomNumber` will pass this test. On the other hand, in theory this unit-test may occasionally fail (for a good enough random number generator). But that is a universal challenge for creating a good unit test for random number generators. Why don't you look-up for an existing unit test on the internet and adapt it to your function?

mgautierfr commented:
I've fixed your comments and add a better test on the `randomNumber` function.

But, honestly I think it is pretty hard to test randomness. I've added the tests because you ask for them.
I haven't found a existing unit-test.
Best comments on internet said that unit test is not a good test for random function.
Some others say we should test the distribution of the results (but it is not enough).

Here I'm felling as testing the std random method. It breaks the black-box principle of writing test but the only reason the output of `randomNumber` is not properly distribute is that std random is broken.
The random number is just use to get a random page. There is nothing critical in the use (especially security).

A careful review would be better than trying to do a complete unit test on a random method.

veloman-yunkan commented on test/random.cpp:
```diff
@@ -36,17 +36,30 @@ TEST(Random, smallMax)
+
+  for (auto i=0U; i<NB_NUMBERS; i++) {
+    auto r = randomNumber(MAX_RANDOM);
+    auto bucket_index = (float)r / MAX_RANDOM * NB_BUCKETS;
+    distribution[bucket_index]++;
```
`bucket_index` can be equal to `NB_BUCKETS`

veloman-yunkan commented on src/fileimpl.h:
```diff
@@ -112,9 +112,9 @@ namespace zim
       entry_index_t getStartUserEntry() const { return m_startUserEntry; }
       entry_index_t getEndUserEntry() const { return m_endUserEntry; }
       // The number of entry added by the creator. (So excluding index, ...).
```
On this line entry was not converted to plural

veloman-yunkan commented on include/zim/archive.h:
```diff
@@ -218,7 +218,7 @@ namespace zim
        * The entry is pick randomly from the front artice list.
```
pick → picked

veloman-yunkan commented on src/archive.cpp:
```diff
@@ -188,7 +188,7 @@ namespace zim
-      auto idx = randomNumber(frontEntryCount);
+      auto idx = randomNumber(frontEntryCount-1);
       auto entry =  getEntryByTitle(idx);
       auto item = entry.getItem(true);
```
The (not so important) corner case of an empty ZIM archive (or, rather, a ZIM archive without any front entries) remained unhandled.

veloman-yunkan reviewed (changes requested):
I was about to approve this PR, and only listed some more minor issues because of a more or less serious issue with `NB_BUCKETS` in the `Random.distribution` unit test.

kelson42 commented:
@mgautierfr This PR is clearly the next one in the pipe, it seems there a few things which still need to be fixed.

mgautierfr commented:
Fix issues found by @veloman-yunkan.
Add better message if random test is failing (which may not be a bug).

mgautierfr commented:
I've removed small.zim from the random unit test as it tends to fail a bit too much.
